modules/tts/tts-adapters/tts-adapter-cereproc/src/main/java/org/daisy/pipeline/tts/cereproc/impl/util/UCharReplacer.java

```diff
@@ -0,0 +1,201 @@
+ * the characteristic 0x-prefix!) of a unicode character as the <tt>key</tt> 
+ * attribute and the replacement string as value of the <tt>entry</tt> 
+ * element.</p>
+ * 
+ * <p>If the <tt>key</tt> attribute contains exactly one unicode codepoint 
+ * (one character) it will be treated literally. It will not be interpreted 
+ * as a HEX representation of another character, even if theoretically 
+ * possible. E.g. if the <tt>key</tt> is "a", it will be
+ * treated as 0x0061 rather than as 0x000a</p>
+ * 
+ * <p>Note - there is a significant difference between a unicode codepoint (32 bit int)
+ * and a UTF16 codeunit (=char) - a codepoint consists of one or two codeunits.</p>  
+ * <p>To make sure an int represents a codepoint and not a codeunit, use for example
+ * <code>com.ibm.icu.text.Normalizer2</code> With Nfc mode.
+ * <code>com.ibm.icu.text.UCharacterIterator</code> to retrieve possibly non-BMP codepoints 
+ * from a string.</p>
+ *  
+ * [1] http://java.sun.com/j2se/1.5.0/docs/api/java/util/Properties.html
+ * [2] http://java.sun.com/dtd/properties.dtd
+ *
+ * Class has been migrated from Pipeline 1
+ *
+ *
+ * @author Markus Gylling
+ * @author carl walinder
+ *
+ */
+
+public class UCharReplacer  {
+	private ArrayList<Map<Integer,String>> mSubstitutionTables; 					//ArrayList<HashMap:<int codepoint>,<replaceString>>: all loaded translationtables
+	private Map<Integer,String> mSubstitutionTable = null;						//represents the currently selected translationTable
+	private Iterator<Map<Integer,String>> mSubstitutionTablesIterator = null;			//recycled tables map iterator
+	private Normalizer2 normalizer;
+	private final static Logger logger = LoggerFactory.getLogger(UCharReplacer.class);
+
+	/**
+	 * Default constructor.
+	 */
+	public UCharReplacer() {
+		mSubstitutionTables = new ArrayList<Map<Integer,String>>();
+		this.normalizer = Normalizer2.getNFDInstance();
+
+	}
+
+	/**
+	 * Add a table for use as source for character substitution strings.	 
+	 * <p>This method can be called multiple times prior to a getReplacementString() 
+	 * call to add several tables. The  table preference order is determined by add 
+	 * order (first added, highest preference).</p>
+	 * 
+	 * @param table URL of the table to add	 
+	 * @throws IOException if the table was not successfully loaded.
+	 */
+	public void addSubstitutionTable(URL table) throws IOException {
+		try{
+			mSubstitutionTables.add(loadTable(table));
+		} catch (Exception e) {
+			throw new IOException(e.getMessage());
+		}
+	}
+	
+	/**
+	 * Replace codepoints in a character sequence with substitute characters.	 
+	 * @param input the character sequence to perform codepoint-to-replacementchars substitution in.
+	 * @return a character sequence whose codepoints may or may not
+	 * have been replaced by substitution characters, depending on settings and substitution success.
+	 */
+	public CharSequence replace(CharSequence input) {
```

Comment:
Can you elaborate, what is not thread safe in that function?

Reply:
I'm seeing at least one issue and that is that `mSubstitutionTable` and `mSubstitutionTablesIterator` are global variables. But there could be more issues, I don't know. You need to check. E.g. I don't know about the UCI classes.

Reply:
Carl, just add the word "synchronized" to all methods that access the members of the class, and the entire class will be thread safe. This method in particular doesn't need that though, since it doesn't access any member directly. It calls the method getSubstitutionChars(), which in turn needs to be synchronized like this:

`private synchronized CharSequence getSubstitutionChars(int codePoint) {	`

Reply:
@carl-textalk It can go in, but only the `retrieveSubstituteFromUserTables` method really needed a "synchronized" keyword. (Possibly `addSubstitutionTable` too however you can assume that this will only be called during initialization of the UCharReplacer object.) Personally I would have just moved `mSubstitutionTable` and `mSubstitutionTablesIterator` inside `retrieveSubstituteFromUserTables` though.

Reply:
Okey, i thought that it would cause overhead to load the substitution-table on every sentence?

Reply:
The table is loaded only once (when `addSubstitutionTable` is called).

The global variables just have no use. They are reassigned in every `retrieveSubstituteFromUserTables` call. There is no overhead brought by making global variables local.

Reply:
Ok then i will do as you suggested then.

Reply:
Ok.